# Day 1 Exercises
## Data Types

#### 1. What types are involved in the following expressions? What data type will each expression evaluate to?


In [ ]:
8 * 2.5

20.0

8 is an integer and 2.5 is a float. This expression will return a float.

In [ ]:
9 / 2

4.5

Both 9 and 2 are integers, this expression will return a float.

In [ ]:
1.5 * 2 >= 7 - 3

False

This expression involves integers and floats. It will return a boolean (true of false).

#### 2. In which order will the expressions be evaluated?

In [ ]:
6 * 3 + 7 * 4

46

First, Python evaluates the multiplications, 6 times 3 and 7 times 4. Then, it adds the results of these together (18+28 = 46).

In [ ]:
5 - 2 * 3 ** 4

-157

First, Python evaluates 3 to the power of 4 (81). Then, this value get's multiplied by 2 (162). Lastly, the resulted value is subtracted from 5 (5-162 = -157).

In [ ]:
(5 - 2) * 3 ** 4

243

Python evaluates 5 minus 2 and 3 to the power of 4 separately. Then, it multiplies these results together (3 x 81 = 243).

In [ ]:
5 + 2 >= 3 * 4

False

Python evaluates 5 plus 2 and 3 times 4 separately. Then evaluates whether 5 plus 2 (7) is greater than or equal to 3 times 4 (12) to return True or False. In this case, since 7 is smaller than 12, it will return False.

## Comments and Errors

#### 1. Which of these expressions results in an error? Why does each error occur?

In [ ]:
((((5 * 4 ** 7))))

81920

No error

In [ ]:
84 *= 0.5 / 7

SyntaxError: 'literal' is an illegal expression for augmented assignment (868797654.py, line 1)

This expression results in an error because *= is not a valid operator.

In [ ]:
(-(-(-(-5 * (4 + 3)))))

35

No error

In [ ]:
5 * 3 = weight

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (3700621706.py, line 1)

This expression results in an error because = is not a valid boolean operator. To evaluate whether these expressions are equal, we should use ==, double equal signs.

In [ ]:
73 / -----------5

-14.6

No error

#### 2. Write a block of code to check if the value is a string. If it is, print out a message saying it is valid. Otherwise, raise an exception with a message saying the value must be a string.

In [ ]:
value = 6
# Your code goes here
if type(value) == str:
  print("The value is valid")
else:
  print("The value must be a string")

The value must be a string


#### 3. What are the following errors?

In [ ]:
arr = [1, 2, 3]
print(arr[3])

IndexError: list index out of range

This error happens because the array doesn't have an index 3. Although it has 3 elements, the indeces start from 0, so we get indeces (0, 1, 2) on each element of the array; rather than (1, 2, 3). As a result, we cannot retrieve the *third* element, using index 3. Instead, we should use print(arr[2]) if we wanted to print the last element of the array.

In [ ]:
fullName = "Taylor Swift"
print(full_name)

NameError: name 'full_name' is not defined

Variable names are case sensitive and must be called exactly as they are created. In this case, the defined variable is called fullName, rather than full_name, as we call within print(). Since full_name is not defined, we get an error when trying to print it.

In [ ]:
2 + '2' + 4

TypeError: unsupported operand type(s) for +: 'int' and 'str'

We cannot use the addition operator across different data types as this will result in an error as shown above.

## Functions

#### 1. Following the function design recipe, define a function that converts kilometers into miles. (Assume there are 1.6 kilometers in a mile.)

In [ ]:
def km_to_mi(km):
  """
  Convert kilometers into miles.
  Parameter: km (number of kilometers, int or float)
  Return: corresponding number of miles (float)
  """
  return km/1.6

#### 2. What value is printed in the below code?

In [ ]:
answer = 3

def answer_to_everything():
    '''Return the answer to life, the universe, and everything'''
    answer = 42
    return answer

answer_to_everything()

print(answer)

3


#### 3. Complete the body of the function. You can assume num is always >= 0.

In [ ]:
def repeat(string, num):
    ''' Return string repeated num times.
    >>> repeat('yes', 4)
    'yesyesyesyes'
    >>> repeat('no', 0)
    ''
    >>> repeat('yesnomaybe', 3)
    'yesnomaybeyesnomaybeyesnomaybe'
    '''
    output = string * num
    return output

#### 4. Complete the function below. You can assume that number will always be a string with 10 digits and that the first 3 digits will always be the area code.

In [ ]:
def is_416_number(number):
    ''' Check whether the number has a 416 area code.
    >>> is_416_number('(416)-555-5555')
    True
    >>> is_416_number('514 416 5555')
    False
    >>> is_416_number('4165554160')
    True
    '''
    # Your code goes here
    # Hint: replace () and spaces with nothing
    number = number.replace('(', '')
    number = number.replace(')', '')
    number = number.replace(' ', '')
    number = number.replace('-', '')
    area_code = int(number[0:3])
    return area_code == 416

# Check:
is_416_number('(416)-555-5555')

True

In [ ]:
# Check
is_416_number('514 416 5555')

False

In [ ]:
# Check
is_416_number('4165554160')

True